packages/beacon-state-transition/src/util/balance.ts

```diff
@@ -52,18 +50,22 @@ export function decreaseBalance(
 }
 
 /**
- * TODO - PERFORMANCE WARNING - NAIVE CODE
  * This method is used to get justified balances from a justified state.
- *
- * SLOW CODE - 🐢
  */
 export function getEffectiveBalances(justifiedState: CachedBeaconState<allForks.BeaconState>): number[] {
-  const justifiedEpoch = justifiedState.currentShuffling.epoch;
-  const effectiveBalances: number[] = [];
-  const validators = readonlyValuesListOfLeafNodeStruct(justifiedState.validators);
-  for (let i = 0, len = validators.length; i < len; i++) {
-    const validator = validators[i];
-    effectiveBalances.push(isActiveValidator(validator, justifiedEpoch) ? validator.effectiveBalance : 0);
+  const {activeIndices} = justifiedState.currentShuffling;
+  // 5x faster than using readonlyValuesListOfLeafNodeStruct
+  const effectiveBalances = justifiedState.effectiveBalances.toArray();
+  let j = 0;
+  for (let i = 0; i < effectiveBalances.length; i++) {
```

Comment:
If you look at the benchmarks

```
MutableVector 250000 iterate all - toArray() + loop | 8.0552 ms/op | 8.9640 ms/op | 0.90
MutableVector 250000 iterate all - get(i) | 3.4577 ms/op | 3.8777 ms/op | 0.89
```
it's actually cheaper to iterate once with individual get(i) ops. Try creating an empty array and push after each get(i)

Reply:
but we need the number[] in the end, it's convenient to create this way and just update to 0 if it's not an active validator

Reply:
also this needs to address https://github.com/ChainSafe/lodestar/issues/2747#issuecomment-879568043

Reply:
@dapplion in the end I switched to use `persistentTS.get(i)`